### ISSUE TITLE
not using all CPU

### ISSUE THREAD

**Issue by lschneidpro** [optuna/optuna#497]
Hi there,

I'm running a ML tuning with optuna on an instance with 48 cores and has set `n_jobs=-1`:
```
study.optimize(OptunaObjective(clf=self.clf,
                                       X=X,
                                       y=y,
                                       sample_weight=sample_weight,
                                       scoring=self.scoring,
                                       cv=self.cv,
                                       decay_f=decay_f),
                       self.n_trials, n_jobs=-1)
```
I noticed that cpu utilization is around 20% and ram is around 4%.

I'm wondering if i set it up correctly or is there is another way to increase performance. Regarding the "Distributed Optimization tutorial", I do not know if i can set it up an sql database as i'm using instances on my university cluster.

If you need more code, let me know. Thank you for any help.

Best regards

**Comment by sile** [comment]
Thank you for using Optuna.

By specifying `n_jobs=-1`, Optuna spawns multiple threads to execute the optimization. However, Python's [GIL](https://wiki.python.org/moin/GlobalInterpreterLock) (Global Interpreter Lock) prevents those threads to run in parallel. Thus, typically, you can only parallelize I/O heavy workload by specifying `n_jobs` option.

The easiest way to utilize your CPU cores is to use [Distributed Optimization](https://optuna.readthedocs.io/en/stable/tutorial/distributed.html) feature on a single machine. 
For this purpose, you can use SQLite as the storage. SQLite is an embedded database (i.e., don't require an external server process) and contained in the Python standard library (https://docs.python.org/3/library/sqlite3.html).

**Comment by lschneidpro** [comment]
thus is this fine in a script? 
```
        study = optuna.create_study(direction='maximize',
                                    study_name='distributed-example',
                                    storage='sqlite:///example.db')
        study.optimize(OptunaObjective(clf=self.clf,
                                       X=X,
                                       y=y,
                                       sample_weight=sample_weight,
                                       scoring=self.scoring,
                                       cv=self.cv,
                                       decay_f=decay_f),
                       self.n_trials, n_jobs=-1)
```
or do i have to keep calling the script from different terminals as suggested in the tutorial ? This anoys me as the optimization is part of a larger pipeline script

**Comment by sile** [comment]
You need to run multiple processes instead of multiple threads, so `n_jobs=-1` option is meaningless for this purpose.

> do i have to keep calling the script from different terminals as suggested in the tutorial ? 

Yes, it's a typical way to run optimization in parallel.
If you don't want to use shell, you can spawn multiple optimization processes within Python script by using a process management library such as `joblib` (please see [quadratic_joblib_simple.py](https://gist.github.com/toshihikoyanase/593246a6d00d6e9ec52de7b7c0f346f3) for an example).

**Comment by lschneidpro** [comment]
Hi Sile,

thanks for your input, it worked. I did something like that:

```
def optuna_optimize(n_trials, study_name, clf, X, y, sample_weight, scoring,
                    cv, decay_f):
    study = optuna.load_study(study_name=study_name,
                              storage='sqlite:///='+str(name)+'.db')

    study.optimize(OptunaObjective(
            clf=clf, X=X, y=y, sample_weight=sample_weight, scoring=scoring,
            cv=cv, decay_f=decay_f), n_trials=n_trials)

    pass
```

```
class Modelling():

    def __init__(self, clf, scoring, cv, n_trials=100, name=None):
    ...
    def train(self, X, y, sample_weight, decay_f=None):

        study = optuna.create_study(direction='maximize',
                                    study_name=self.name,
                                    storage='sqlite:///='+str(self.name)+'.db',
                                    load_if_exists=True)

        r = Parallel(n_jobs=-1)([delayed(optuna_optimize)(
                self.n_trials,
                self.name,
                self.clf,
                X,
                y,
                sample_weight,
                self.scoring,
                self.cv,
                decay_f) for i in tqdm(range(self.n_trials))])
   ...
```


On 48 cores, it quickly crashed with:
`sqlite3.OperationalError: database is locked`

I assume I have to move to MySQL our Postgres. I asked my university cluster support to know if i can run locally such a DB.

Otherwise I had an issue regarding pruning with LightGBM, but i'll raise another issue I guess :)

**Comment by sile** [comment]
> On 48 cores, it quickly crashed with:
> `sqlite3.OperationalError: database is locked`
> I had to drop to 6 cores to run it without any issues.

Hmm...
Thank you for telling me the error. SQLite seems to have a scalability problem.

> I assume I have to move to MySQL our Postgres. I asked my university cluster support to know if i can run locally such a DB.

Sorry for the inconvenience. I think that using MySQL or PostgreSQL is a sure way.

> Otherwise I had an issue regarding pruning with LightGBM, but i'll raise another issue I guess :)

:)

---

By the way,  your code seems to run `self.n_trials**2` trials, so it may be better to modify as follows:

```python
        r = Parallel(n_jobs=-1)([delayed(optuna_optimize)(
                1,  # FIXED (original: `self.n_trials`)
                self.name,
                self.clf,
                X,
                y,
                sample_weight,
                self.scoring,
                self.cv,
                decay_f) for i in tqdm(range(self.n_trials))])
```

**Comment by lschneidpro** [comment]
Thanks again for spoting such a mistake.
It seems that the author of your example is running `10**2 trials` then
> Yes, it's a typical way to run optimization in parallel.
> If you don't want to use shell, you can spawn multiple optimization processes within Python script by using a process management library such as `joblib` (please see [quadratic_joblib_simple.py](https://gist.github.com/toshihikoyanase/593246a6d00d6e9ec52de7b7c0f346f3) for an example).

**Comment by lschneidpro** [comment]
Ok my uni cluster support came back to me. Running such a DB would have be done externally by an another IT department. So it cannot be guaranteed to work.

Could I use such a trick instead: https://github.com/pfnet/optuna/issues/423 ?

**Comment by sile** [comment]
> Could I use such a trick instead: #423 ?

You can change SQLite (SQLAlchemy) options in the way shown in #423.
For instance, the following code increases timeout seconds and it may reduce the probability of "database is locked" error.
(I'm not sure this is the answer you expected, though)
```python
sqlite_timeout = 300
storage = optuna.storages.RDBStorage(
               'sqlite:///='+str(self.name)+'.db',
                engine_kwargs={'connect_args': {'timeout': sqlite_timeout}}
            )
study = optuna.create_study(storage=storage)
```

---

BTW, I tried implementing [`MultiProcessStorage`](https://gist.github.com/sile/2c329d0cb38f35e25a8c7bab44edbb63#file-multi_process_storage-py)  that can be used to parallelize optimization without RDB (the storage uses `multiprocessing` module internally).
An usage example of this storage is found in [the gist](https://gist.github.com/sile/2c329d0cb38f35e25a8c7bab44edbb63#file-quadratic_simple-py
).
This is a highly experimental storage, but it seemed to work as expected in my environment.

**Comment by lschneidpro** [comment]
I could try.

In the mean time, I started renting a Postgre SQL server online (30 first days free)

**Comment by sile** [comment]
Closing this issue since there are no activities for several months. Feel free to reopen or create a new issue if you have any problems.

**Comment by rikkigouda** [comment]
Hi there! 👋 I'm seemingly facing the same issue in GCP running optuna on an Ubuntu instance.

Any suggestion on how to get around it? The CPU utilization is capped at 50% and isn't using the full resource.

Thanks!
